### PR TITLE
feat(a11y): enhance keyboard interaction in search mode

### PR DESCRIPTION
**Pull request by aojunhao123**
## 🚀 Feature Description

Enhanced keyboard interaction in search mode to improve accessibility.

### Changes
- Allow keyboard users to select first matched non-disabled node by pressing enter
- Prevent selecting unrelated active node when search value exists
- Ignore enter key press when no matches found or all matched nodes are disabled

### 🎯 Motivation
Improve accessibility for keyboard users by providing a more intuitive way to select nodes when searching. Previously, pressing enter would select the active node regardless of search context, which could lead to unexpected selections.

### 📝 Implementation Details
Modified the keyboard event handling in `OptionList.tsx` to:
1. Check for search value existence when enter key is pressed
2. Find and select the first non-disabled node that matches the search criteria
3. Maintain existing keyboard navigation behavior when no search is active

### 🧪 Testing
Added comprehensive test cases in `Select.SearchInput.spec.js` covering:
- Selecting first matched node on enter
- Handling disabled nodes
- Edge cases with no matches
- Interaction with search value

### 📋 Checklist
- [x] Added/updated tests
- [x] All tests passing
- [x] No breaking changes
- [x] Follows accessibility best practices

### 📸 Screenshots/GIFs
N/A

### 📚 Related Issues
- fix https://github.com/ant-design/ant-design/issues/50852

## Summary by CodeRabbit

- **新功能**
	- 改进了`OptionList`组件的搜索功能，新增了根据搜索条件识别首个可选节点的方法。
	- 更新了组件的活动键管理逻辑，确保在打开时优先激活匹配节点。
	- 增强了组件的可访问性，确保在打开下拉菜单时自动激活首个可选项。
- **错误修复**
	- 防止选择禁用节点，确保只有可选节点可以被激活。
- **测试**
	- 增强了组件的键盘事件处理测试，确保在输入有效搜索词后按下Enter键能够正确选择节点。
	- 添加了快照测试，以确保组件在不同状态下的渲染输出一致。
	- 扩展了事件处理测试，确保组件在用户交互时表现正常。
- **杂项**
	- 更新了`.gitignore`文件，以忽略更多的配置文件和锁文件。

### CODE REVIEW

File: src/OptionList.tsx [hunk 1]
```diff
@@ -175,7 +190,18 @@ const OptionList: React.ForwardRefRenderFunction<ReviseRefOptionListProps> = (_,
 
         // >>> Select item
         case KeyCode.ENTER: {
-          if (activeEntity) {
+          if (searchValue) {
+            const firstMatchNode = getFirstMatchNode(memoTreeData);
```
Comment (zombieJ) [inline src/OptionList.tsx:194]: 这不能够认为第一个就是，对于视障用户是不知道第一个选项是什么的。如果是要做这个处理，应该是 search 的时候默认 active 第一个。或者说第一次回车 active 第一个，然后第二次回车选中才对。

Comment (aojunhao123) [reply src/OptionList.tsx:194]: get了。站在用户体验的角度考虑的话，search的时候默认active第一个，然后进行回车的选中操作应该要比两次回车要好一点，我晚点重新改一版

File: src/OptionList.tsx [hunk 2]
```diff
@@ -102,6 +102,57 @@ const OptionList: React.ForwardRefRenderFunction<ReviseRefOptionListProps> = (_,
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [open]);
 
+  // =========================== Search Effect ===========================
+  React.useEffect(() => {
+    if (searchValue) {
```
Comment (zombieJ) [inline src/OptionList.tsx:107]: 我发现没有搜索的时候，也是没有高亮第一个的。如果搜索会默认高亮，那其实都可以默认高亮第一个。另外 CI failed

File: src/OptionList.tsx [hunk 3]
```diff
@@ -138,26 +176,35 @@ const OptionList: React.ForwardRefRenderFunction<ReviseRefOptionListProps> = (_,
...
+      setSearchExpandedKeys(getAllKeys(treeData, fieldNames));
 
-    if (!multiple) {
-      toggleOpen(false);
+      const firstMatchNode = getFirstMatchNode(memoTreeData);
+      if (firstMatchNode) {
+        setActiveKey(firstMatchNode[fieldNames.value]);
```
Comment (zombieJ) [inline src/OptionList.tsx:201]: search effect 和 scroll effect 重复了，应该合并成一个。

File: src/OptionList.tsx [hunk 4]
```diff
@@ -138,26 +149,59 @@ const OptionList: React.ForwardRefRenderFunction<ReviseRefOptionListProps> = (_,
...
+    getFirstMatchingNode(nodes, node => node.selectable !== false && !node.disabled);
 
-    if (checkable && isCheckDisabled(node)) {
+  const getFirstMatchNode = (nodes: EventDataNode<any>): EventDataNode<any> | null =>
+    getFirstMatchingNode(nodes, node => filterTreeNode(node) && !node.disabled);
+
+  // ========================== Active Key Effect ==========================
+  React.useEffect(() => {
+    if (searchValue) {
+      const firstMatchNode = getFirstMatchNode(memoTreeData);
```
Comment (zombieJ) [inline src/OptionList.tsx:187]: 有没有 `searchValue` 你的 matchNode 逻辑应该都是一样的。要么都是 selectable 的，要么都是非 disabled 的。可以考虑直接从 virtualList 里获取，这样也不需要递归，并且给出来的 TreeNode 本来就是过滤完的。

File: examples/basic.tsx [hunk 5]
```diff
@@ -216,7 +216,6 @@ class Demo extends React.Component {
             console.log('onPopupScroll:', evt.target);
           }}
         />
-
```
Comment (zombieJ) [inline examples/basic.tsx:219]: 空行不用删，这是为了看 jsx 的时候区分方便哈~

File: src/OptionList.tsx [hunk 6]
```diff
@@ -10,6 +10,7 @@ import LegacyContext from './LegacyContext';
 import TreeSelectContext from './TreeSelectContext';
 import type { Key, SafeKey } from './interface';
 import { getAllKeys, isCheckDisabled } from './utils/valueUtil';
+import { flattenTreeData } from 'rc-tree/lib/utils/treeUtil';
```
Comment (zombieJ) [inline src/OptionList.tsx:13]: 没清掉

File: src/OptionList.tsx [hunk 7]
```diff
@@ -76,7 +77,7 @@ const OptionList: React.ForwardRefRenderFunction<ReviseRefOptionListProps> = (_,
     (prev, next) => next[0] && prev[1] !== next[1],
   );
 
-  // ========================== Active ==========================
+  // ========================== Active Key ==========================
```
Comment (zombieJ) [inline src/OptionList.tsx:80]: 这个注释不用改，`// =====` 是语义化的区域划分，Active 就是管 Active 不是仅仅管 `Key`

File: src/OptionList.tsx [hunk 8]
```diff
@@ -92,16 +93,36 @@ const OptionList: React.ForwardRefRenderFunction<ReviseRefOptionListProps> = (_,
     };
   }, [checkable, checkedKeys, halfCheckedKeys]);
 
-  // ========================== Scroll ==========================
+  // ========================== Scroll Effect ==========================
```
Comment (zombieJ) [inline src/OptionList.tsx:96]: 同上，这是管 Scroll 相关逻辑。不是管 Scroll Effect

File: src/OptionList.tsx [hunk 9]
```diff
@@ -138,26 +152,66 @@ const OptionList: React.ForwardRefRenderFunction<ReviseRefOptionListProps> = (_,
     }
   };
 
-  // ========================== Events ==========================
-  const onListMouseDown: React.MouseEventHandler<HTMLDivElement> = event => {
-    event.preventDefault();
-  };
+  // ========================== Search Effect ==========================
```
Comment (zombieJ) [inline src/OptionList.tsx:155]: 看起来位置不用挪，这就是 Search 的内容，不用单独划区 `// ====`

File: src/OptionList.tsx [hunk 10]
```diff
@@ -138,26 +152,66 @@ const OptionList: React.ForwardRefRenderFunction<ReviseRefOptionListProps> = (_,
...
+    nodes: EventDataNode<any>[],
+    searchVal?: string,
+  ): EventDataNode<any> | null => {
+    for (const node of nodes) {
+      if (node.disabled || node.selectable === false) {
+        continue;
+      }
 
-    if (checkable && isCheckDisabled(node)) {
+      if (searchVal) {
```
Comment (zombieJ) [inline src/OptionList.tsx:173]: 不用根据 searchVal 和 filterTreeNode 来分叉，OptionList 拿到的 treeData 本来就是过滤后的节点:

https://github.com/react-component/tree-select/blob/fb1b6bb6fc2d4ae2cdd47f841a4485fa41a0f019/src/TreeSelect.tsx#L606

File: src/OptionList.tsx [hunk 11]
```diff
@@ -138,26 +152,66 @@ const OptionList: React.ForwardRefRenderFunction<ReviseRefOptionListProps> = (_,
...
+        }
+      }
+    }
+    return null;
+  };
+
+  // ========================== Active Key Effect ==========================
+  React.useEffect(() => {
+    if (!open) {
+      setActiveKey(null);
```
Comment (zombieJ) [inline src/OptionList.tsx:194]: 这个不需要，在关闭的时候如果为 `null` 会导致动画关闭阶段看到 active 消失了

File: src/OptionList.tsx [hunk 12]
```diff
@@ -138,26 +138,69 @@ const OptionList: React.ForwardRefRenderFunction<ReviseRefOptionListProps> = (_,
...
-    if (checkable && isCheckDisabled(node)) {
+  // ========================== Get First Selectable Node ==========================
+  const getFirstMatchingNode = (nodes: EventDataNode<any>[]): EventDataNode<any> | null => {
+    for (const node of nodes) {
+      if (!node.disabled && node.selectable !== false) {
+        return node;
+      }
+
+      if (node[fieldNames.children]) {
+        const matchInChildren = getFirstMatchingNode(node[fieldNames.children]);
```
Comment (zombieJ) [inline src/OptionList.tsx:165]: 这里如果遇到为空的会挂掉吧？

File: src/OptionList.tsx [hunk 13]
```diff
@@ -138,26 +138,69 @@ const OptionList: React.ForwardRefRenderFunction<ReviseRefOptionListProps> = (_,
...
+      nextActiveKey = getFirstNode();
     }
-  };
+    // single mode active first checked node
+    else if (!multiple && checkedKeys.length) {
+      nextActiveKey = checkedKeys[0];
+    }
+    // default active first node
+    else {
+      nextActiveKey = getFirstNode();
```
Comment (zombieJ) [inline src/OptionList.tsx:199]: 既然 searchValue 和 else 都到这里，那不就是先 `if (!multiple && checkedKeys.length)` 然后剩下的都走 get 不就好了？

**Comment by afc163** [comment]
覆盖率掉了

**Comment by aojunhao123** [comment]
> 覆盖率掉了

我改一下
